pkg/controller/certificaterequests/vault/fuzz_test.go

```diff
@@ -0,0 +1,186 @@
+			gen.SetCertificateRequestIssuer(cmmeta.ObjectReference{
+				Name:  baseIssuer.Name,
+				Group: certmanager.GroupName,
+				Kind:  baseIssuer.Kind,
+			}),
+		)
+		var condition cmapi.CertificateRequestConditionType
+		switch baseCRCondition % 4 {
+		case 0:
+			condition = cmapi.CertificateRequestConditionReady
+		case 1:
+			condition = cmapi.CertificateRequestConditionInvalidRequest
+		case 2:
+			condition = cmapi.CertificateRequestConditionApproved
+		case 3:
+			condition = cmapi.CertificateRequestConditionDenied
+		}
+		baseCR := gen.CertificateRequestFrom(baseCRNotApproved,
+			gen.SetCertificateRequestStatusCondition(cmapi.CertificateRequestCondition{
+				Type:               condition,
+				Status:             cmmeta.ConditionTrue,
+				Reason:             "cert-manager.io",
+				Message:            "Certificate request has been approved by cert-manager.io",
+				LastTransitionTime: &metaFixedClockStart,
+			}),
+		)
+
+		kubeObjects := []runtime.Object{}
+		certManagerObjects := []runtime.Object{}
+		certManagerObjects = append(certManagerObjects, baseCR.DeepCopy())
+
+		// Add token if the fuzzer decides to.
+		if addToken {
+			tokenSecret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: gen.DefaultTestNamespace,
+					Name:      "token-secret",
+				},
+				Data: map[string][]byte{
+					"my-token-key": secretTokenData,
+				},
+			}
+			kubeObjects = append(kubeObjects, tokenSecret)
+			certManagerObjects = append(certManagerObjects, gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerVault(cmapi.VaultIssuer{
+					Auth: cmapi.VaultAuth{
+						TokenSecretRef: &cmmeta.SecretKeySelector{
+							Key: "my-token-key",
+							LocalObjectReference: cmmeta.LocalObjectReference{
+								Name: "token-secret",
+							},
+						},
+					},
+				}),
+			))
+		} else {
+			certManagerObjects = append(certManagerObjects, baseIssuer.DeepCopy())
+		}
+
+		builder := &testpkg.Builder{
+			T:                  t,
+			KubeObjects:        kubeObjects,
+			CertManagerObjects: certManagerObjects,
+		}
+		builder.Init()
+		defer builder.Stop()
+		vault := NewVault(builder.Context).(*Vault)
+
+		if !addCustomCsrPEM {
+			rsaPEMCert, err := generateSelfSignedCertFromCR(baseCR, rsaSKFuzz)
+			if err != nil {
+				return
+			}
+
+			fakeVault := fakevault.New().WithSign(rsaPEMCert, rsaPEMCert, nil)
+			vault.vaultClientBuilder = func(_ context.Context, ns string, _ func(ns string) internalvault.CreateToken, sl internalinformers.SecretLister,
+				iss cmapi.GenericIssuer) (internalvault.Interface, error) {
+				return fakeVault.New(ns, sl, iss)
+			}
+		}
+
+		controller := certificaterequests.New(
+			apiutil.IssuerVault,
+			func(*controllerpkg.Context) certificaterequests.Issuer { return vault },
+		)
+		if _, _, err := controller.Register(builder.Context); err != nil {
+			// Make it explicit if this fails
+			panic(err)
+		}
+		builder.Start()
+		controller.Sync(context.Background(), baseCR)
```

Comment:
What are we testing for here? Are we just trying to find panics?

Reply:
Yep. I will be happy to look into checking for certain conditions, but I would prefer to get it running on OSS-Fuzz first. But yes, currently it checks for panics, OOM's and stack overflows.

Reply:
Could you add that in a code comment? Rest of the PR looks good to me.